apps/server/lib/lexical/server/code_intelligence/completion.ex

```diff
@@ -38,34 +38,30 @@ defmodule Lexical.Server.CodeIntelligence.Completion do
     case Env.new(project, document, position) do
       {:ok, env} ->
         completions = completions(project, env, context)
-        Logger.warning("Emitting completions: #{inspect(completions)}")
-        completion_list(completions)
+        Logger.info("Emitting completions: #{inspect(completions)}")
+        maybe_to_completion_list(completions)
 
       {:error, _} = error ->
         Logger.error("Failed to build completion env #{inspect(error)}")
-        completion_list()
+        maybe_to_completion_list()
     end
   end
 
   defp completions(%Project{} = project, %Env{} = env, %Completion.Context{} = context) do
     prefix_tokens = Env.prefix_tokens(env, 1)
 
     cond do
-      prefix_tokens == [] ->
+      prefix_tokens == [] or not should_emit_completions?(env) ->
         []
 
-      match?([{:operator, :do, _}], prefix_tokens) and Env.empty?(env.suffix) ->
+      should_emit_do_end_snippet?(env) ->
         do_end_snippet = "do\n  $0\nend"
 
         env
         |> Builder.snippet(do_end_snippet, label: "do/end block")
         |> List.wrap()
 
-      Enum.empty?(prefix_tokens) or not context_will_give_meaningful_completions?(env) ->
-        []
-
-      Env.in_context?(env, :struct_arguments) and not Env.in_context?(env, :struct_field_value) and
-          not prefix_is_trigger?(env) ->
+      Env.in_context?(env, :struct_field_key) ->
         project
         |> RemoteControl.Api.complete_struct_fields(env.document, env.position)
         |> Enum.map(&Translatable.translate(&1, Builder, env))
```

Comment:
So our solution to the vscode-completion problem is to have a single character prefix to this branch, right?

<img width="564" alt="CleanShot 2023-10-16 at 14 31 41@2x" src="https://github.com/lexical-lsp/lexical/assets/12830256/80b97d17-971f-4a4f-b7d8-74a69bb4ca20">

Reply:
Yes, we must return completions for single character prefixes in VS Code if a request is issued because, otherwise, no further completions will work for that word.